clients/include/norm.hpp

```diff
@@ -21,28 +28,260 @@
+    host_vector<double> hGPU_double(N * lda);
+
+    for(rocblas_int i = 0; i < N * lda; i++)
+    {
+        hCPU_double[i] = double(half_to_float(hCPU[i]));
+        hGPU_double[i] = double(half_to_float(hGPU[i]));
+    }
+
+    return norm_check_general(norm_type, M, N, lda, hCPU_double.data(), hGPU_double.data());
+}
 
 template <typename T>
-double norm_check_symmetric(
-    char norm_type, char uplo, rocblas_int N, rocblas_int lda, T* hCPU, T* hGPU);
+inline double norm_check_general(char        norm_type,
+                                 rocblas_int M,
+                                 rocblas_int N,
+                                 rocblas_int lda,
+                                 rocblas_int stride_a,
+                                 rocblas_int batch_count,
+                                 T*          hCPU,
+                                 T*          hGPU)
+{
+    // norm type can be O', 'I', 'F', 'o', 'i', 'f' for one, infinity or Frobenius norm
+    // one norm is max column sum
+    // infinity norm is max row sum
+    // Frobenius is l2 norm of matrix entries
+    //
+    // use triangle inequality ||a+b|| <= ||a|| + ||b|| to calculate upper limit for Frobenius norm
+    // of strided batched matrix
+
+    double cumulative_error = 0.0;
+
+    for(rocblas_int i = 0; i < batch_count; i++)
+    {
+        auto index = i * stride_a;
+
+        auto error = norm_check_general(norm_type, M, N, lda, hCPU + index, hGPU + index);
+
+        if(norm_type == 'F' || norm_type == 'f')
+        {
+            cumulative_error += error;
+        }
+        else if(norm_type == 'O' || norm_type == 'o' || norm_type == 'I' || norm_type == 'i')
+        {
+            cumulative_error = cumulative_error > error ? cumulative_error : error;
+        }
+    }
+
+    return cumulative_error;
+}
+
+/* ============== Norm Check for Symmetric Matrix ============= */
+/*! \brief compare the norm error of two hermitian/symmetric matrices hCPU & hGPU */
+
+template <typename T, typename std::enable_if<!is_complex<T>>::type* = nullptr>
+inline double norm_check_symmetric(
+    char norm_type, char uplo, rocblas_int N, rocblas_int lda, T* hCPU, T* hGPU)
+{
+    // norm type can be M', 'I', 'F', 'l': 'F' (Frobenius norm) is used mostly
+
+    double      work[1];
+    rocblas_int incx  = 1;
+    double      alpha = -1.0;
+    rocblas_int size  = lda * N;
+
+    host_vector<double> hCPU_double(N * lda);
+    host_vector<double> hGPU_double(N * lda);
+
+    for(rocblas_int i = 0; i < N * lda; i++)
+    {
+        hCPU_double[i] = double(hCPU[i]);
+        hGPU_double[i] = double(hGPU[i]);
+    }
+
+    double cpu_norm = dlansy_(&norm_type, &uplo, &N, hCPU_double, &lda, work);
+    daxpy_(&size, &alpha, hCPU_double, &incx, hGPU_double, &incx);
+
+    double error = dlansy_(&norm_type, &uplo, &N, hGPU_double, &lda, work) / cpu_norm;
+
+    return error;
+}
+
+template <typename T, typename std::enable_if<is_complex<T>>::type* = nullptr>
```

Comment:
As above:
`template <typename T, typename std::enable_if<!is_complex<T>, int>::type = 0>`

Reply:
understand. update it in commit "update complex change 2".